r2-core/src/main/java/com/linkedin/r2/transport/http/client/SmoothRateLimiter.java

```diff
@@ -317,12 +318,12 @@ public void loop()
         try
         {
           // avoids executing too many duplicate tasks
-          long nextRunRelativeTime = Math.max(0, _permitTime + rate.getPeriod() - _clock.currentTimeMillis());
-          long nextRunAbsolute = _clock.currentTimeMillis() + nextRunRelativeTime;
-          if (_nextScheduled > nextRunAbsolute || _nextScheduled <= _clock.currentTimeMillis())
+          // reschedule next iteration of the event loop to the next delay, or the beginning of the next period
+          long nextRunRelativeTime = _permitAvailableCount > 0 ? _delayUntil - now : Math.max(0, _permitTime + rate.getPeriod() - now);
+          long nextRunAbsolute = now + nextRunRelativeTime;
```

Comment:
This looks a bit confusing to me. How about keep nextRunRelativeTime as is and change
`long nextRunAbsolute = _permitAvailableCount > 0 ? _delayUntil : now + nextRunRelativeTime`

Reply:
The suggested change results in EventLoop failing to honor the delay, with all unit tests failing in TestConstantQpsRateLimiter, though all tests in TestSmoothRateLimiter continue to pass. We need to augment both nextRunRelative and nextRunAbsolute with the delay in order for the next iteration of the loop to be scheduled with the delay in mind.